### PR TITLE
Refactoring encoding related code and resolve related FIXMEs

### PR DESCRIPTION
**Pull request by huansong**
This PR attempts to refactor the code around processing the encoding clauses mainly in CREATE TABLE, ALTER TABLE, CREATE TYPE and ALTER TYPE. A total of 15 related FIXMEs are removed here.

Also removed another FIXME in `makeObjectName()` that's not related to encoding but it's fairly straightforward: we added an elog(ERROR) to workaround an assert failure in the regress test. But I ran the regress test and didn't see the failure, so restored the code and removed the FIXME. If there's ever an assert failure again, we should fix the same instead of adding another FIXME.

The changes mainly include:
* Refactor encoding code for CREATE TABLE and ALTER TABLE 
  * Remove `transformAttributeEncoding()` and create `transformColumnEncoding()` to be used in both `DefineRelation()` and `ATExecAddColumn()`
  * Move encoding related code in `ATPrepAddColumn()` to `ATExecAddColumn()` so they are in one place.
  * In `transformColumnEncoding()`, correct the code to match the comment `If no default has been specified, we might create one out of the WITH clause.`
* Refactor encoding code for CREATE TYPE and ALTER TYPE
  * Simplify and unify encoding code in `DefineType()` and `AlterType()`. Move the utility functions to pg_type.c
  * After refactoring, the code should be straightforward enough (`transformStorageEncodingClause()` and then `transformRelOptions()`), so I think we don't need to over-simplify it by creating a single function interface as indicated by the FIXME.
* Refactor `prebuild_temp_table()`
  * Simplify the code. Didn't create new function as well since the current form should be good enough. The encoding related logic is not "leaked" anymore.
* Move functions to more appropriate source files
  * `rel_get_column_encodings()`, `is_storage_encoding_directive()`, `default_column_encoding_clause()`, `remove_type_encoding()`, `TypeNameGetStorageDirective()`

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/backend/commands/tablecmds.c [hunk 1]
```diff
@@ -7386,45 +7348,24 @@ ATExecAddColumn(List **wqueue, AlteredTableInfo *tab, Relation rel,
...
-
 	/* 
+	 * Process the encoding clauses.
+	 *
 	 * For AO/CO tables, always store an encoding clause. If no encoding
 	 * clause was provided, store the default encoding clause.
+	 * If there's an encoding clause for non AO/CO tables, we'll throw an error 
+	 * in the function (indicated by errorOnEncodingClause == true).
+	 */
+	List* enc = transformColumnEncoding(rel, list_make1(colDef),
```
Comment (jimmyyih) [inline src/backend/commands/tablecmds.c:7359]: The declaration of `enc` needs to be at the top of the scope (in this case, at the top of the function) as per Postgres coding guidelines.

**Review by jimmyyih — APPROVED** [review]
Nice refactor.

File: src/backend/commands/tablecmds.c (hunk 1 above)
Comment (huansong) [reply src/backend/commands/tablecmds.c:7359]: Ack. Done.

File: src/backend/access/common/reloptions_gp.c [hunk 2]
```diff
@@ -1503,37 +1577,26 @@ find_crsd(const char *column, List *stenc)
...
- *		but not consistently in all cases. Look at CREATE and ALTER table for
- *		specifics.
  */
-List *
-transformAttributeEncoding(List *columns,
-						   List *stenc,
-						   List *withOptions,
-						   bool rootpartition,
-						   bool *found_enc)
+List* transformColumnEncoding(Relation rel, List *colDefs, List *stenc, List *withOptions, bool rootpartition, bool errorOnEncodingClause)
```
[the remark this answers is not in the document]
Comment (huansong) [reply src/backend/access/common/reloptions_gp.c:1581]: Done. Also adjusted a few places in the comment of `transformColumnEncoding()`.

**Review by kainwen — COMMENTED** [review]
Reading parts, not finished yet.

File: src/backend/access/common/reloptions_gp.c [hunk 3]
```diff
@@ -1202,6 +1223,59 @@ fillin_encoding(List *aocoColumnEncoding)
...
+	{
+		GetAppendOnlyEntryAttributes(RelationGetRelid(rel),
+									 &blocksize,
+									 NULL,
+									 &compresslevel,
+									 NULL,
+									 &compresstype_nd);
+		compresstype = NameStr(compresstype_nd);
+	}
+
```
[the remark this answers is not in the document]
Comment (huansong) [reply src/backend/access/common/reloptions_gp.c:1254]: Done. I'm re-using `compresstype` just not to over-complicate it. 

Btw just a side note that this PR mainly addresses the FIXMEs so I guess we don't need to focus too much on other aesthetic issues (but if we see any obvious ones we should definitely fix them).

File: src/backend/access/common/reloptions_gp.c [hunk 4]
```diff
@@ -1088,6 +1088,27 @@ get_option_set(relopt_value *options, int num_options, const char *opt_name)
  * ------------------------------------------------------------------------
  */
 
+/* names we expect to see in ENCODING clauses */
```
[the remark this answers is not in the document]
Comment (huansong) [reply src/backend/access/common/reloptions_gp.c]: Yes that's better. Done.

**Pull request merged**